src/vm/analysis/trait_checker/mod.rs

```diff
@@ -0,0 +1,78 @@
+use std::collections::HashMap;
+
+use vm::analysis::types::{ContractAnalysis, AnalysisPass};
+use vm::analysis::AnalysisDatabase;
+use vm::analysis::errors::{CheckResult, CheckError, CheckErrors};
+use vm::representations::{SymbolicExpression, ClarityName};
+use vm::representations::SymbolicExpressionType::{AtomValue, Atom, List, LiteralValue};
+use vm::types::{Value, TraitIdentifier, TypeSignature, FunctionType};
+use vm::functions::NativeFunctions;
+use vm::functions::define::{DefineFunctions, DefineFunctionsParsed};
+
+pub struct TraitChecker {
+}
+
+impl AnalysisPass for TraitChecker {
+
+    fn run_pass(contract_analysis: &mut ContractAnalysis, analysis_db: &mut AnalysisDatabase) -> CheckResult<()> {
+        let mut command = TraitChecker::new();
+        command.run(contract_analysis, analysis_db)?;
+        Ok(())
+    }
+}
+
+impl TraitChecker {
+
+    fn new() -> Self {
+        Self {
+        }
+    }
+
+    pub fn run(&mut self, contract_analysis: &mut ContractAnalysis, analysis_db: &mut AnalysisDatabase) -> CheckResult<()> {
+    
+        for trait_identifier in &contract_analysis.implemented_traits {
+
+            let trait_name = trait_identifier.name.to_string();
+            let contract_defining_trait = analysis_db.load_contract(&trait_identifier.contract_identifier)
+                .ok_or(CheckErrors::TraitReferenceUnknown(trait_identifier.name.to_string()))?;
+            let trait_sig = contract_defining_trait.get_defined_trait(&trait_name)
+                .ok_or(CheckErrors::TraitReferenceUnknown(trait_identifier.name.to_string()))?;
+
+            for (func_name, expected_sig) in trait_sig.iter() {
+                match contract_analysis.get_public_function_type(func_name) {
+                    Some(FunctionType::Fixed(func)) => {
+                        if func.args.len() != expected_sig.args.len() {
+                            return Err(CheckErrors::BadTraitImplementation(trait_name.clone(), func_name.to_string()).into())
+                        }
+                        let args = expected_sig.args.iter().zip(func.args.iter());
+                        for (expected_arg, arg) in args {
+                            match (expected_arg, &arg.signature) {
+                                (TypeSignature::TraitReferenceType(expected), TypeSignature::TraitReferenceType(actual)) => {
+                                    if actual != expected {
+                                        return Err(CheckErrors::BadTraitImplementation(trait_name.clone(), func_name.to_string()).into())
+                                    }
+                                }
+                                _ => {
+                                    if !expected_arg.admits_type(&arg.signature) {
```

Comment:
This should either be:

```
if !arg.signature.admits_type(&expected_arg) {
    ...
```

(That makes the implementing function arg _more_ permissive than the impl definition)

The problem with the check as it currently implemented can be seen with something like:


```
(define-trait trait-1 (
            (get-1 ((list 10 uint)) (response uint uint))))
```

Consider the following attempts to implement:
```
(define-public (get-1 (list 5 uint)) 
            (contract-call? contract get-1 u0))";
```
This should fail -- but I think would pass a check like `expected_arg.admits_type(&arg.signature)`.

```
(define-public (get-1 (list 15 uint)) 
            (contract-call? contract get-1 u0))";
```

This is actually okay --- since any caller can be sure that passing a `(list 10 uint)` would not result in a type error.

For return types, the directionality of admission is implemented correctly (the implementor's return type should be less permissive than the expected return type).

Reply:
Addressed with https://github.com/blockstack/blockstack-core/pull/1246/commits/058549a72fd3e882bcef0b45ccd31c12a55051ee, Test cases added with https://github.com/blockstack/blockstack-core/pull/1246/commits/d409a83f4265a5bac8ad3ae19705c2d38d2f4a87